src/tools/rbd_mirror/image_map/ImageAction.h

```diff
@@ -0,0 +1,138 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_IMAGE_ACTION_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_IMAGE_ACTION_H
+
+#include <map>
+#include <list>
+#include <string>
+
+#include "Types.h"
+#include "common/Mutex.h"
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+class ImageAction {
+public:
+  ImageAction();
+
+  enum MirrorPeerType {
```

Comment:
You only need to track actions at the global image id level -- you shouldn't care about state transitions for individual mirror peer images. That is why I have a comments below that you should track images using a "global image id -> set<mirror uuid>" mapping, that you should map an image once it is first added, and you should unmap and image once all mirror peers (remote or local) have been removed from said mapping.